Data seeding / Server setup (Kaz)

ahchung23 opened the pull request:
**### Features:**

**-seeds database with database/seed.js**
database schema described in schema.sql
seed.js uses methods described in seedMethods.js
dataObj created in sampleDataModel.js
functions to create dataObj described in sampleDataMethods.js

**-routes get and pull requests from client**
endpoints and server routing described in server/index.js.
set static path to "/books/:id" which serves the index.html in public folder.

**### Questions:**

1. Do you think my data schema (in schema.sql) design looks good?
2. Do you see any problems with the data I seeded into my database?
3. Do you see any potential problems with the API endpoints i'm using?
4. What do you think should be the static path to initially serve up an item page? Do you agree with "/books/:id" ?

**_Feedback much appreciated!_** 👍

kazshige commented on schema.sql:
```diff
@@ -0,0 +1,40 @@
+CREATE TABLE IF NOT EXISTS characters (
+  id INT NOT NULL AUTO_INCREMENT PRIMARY KEY,
+  name VARCHAR(100),
+  bookId INT
```
Where is "bookId" coming from? What table does it relate to?

ahchung23 replied:
The bookId refers to the id of the book in the main details table. The details table holds 100 data sets for 100 books.

kazshige commented on schema.sql:
```diff
@@ -0,0 +1,40 @@
+  id INT NOT NULL AUTO_INCREMENT PRIMARY KEY,
+  name VARCHAR(100),
+  year INT,
+  bookId INT
```
Same as above

ahchung23 replied:
The bookId refers to the id of the book in the main details table. The details table holds 100 data sets for 100 books :)

kazshige commented on schema.sql:
```diff
@@ -0,0 +1,40 @@
+  publisher VARCHAR(100),
+  originalPubDate VARCHAR(30),
+  coverurl VARCHAR(250),
+  bookId INT
```
Same as above

ahchung23 replied:
The bookId refers to the id of the book in the main details table. The details table holds 100 data sets for 100 books :)

kazshige commented on schema.sql:
```diff
@@ -0,0 +1,40 @@
+  originalPubDate VARCHAR(30),
+  title VARCHAR(100),
+  isbn10 VARCHAR(20),
+  isbn13 VARCHAR(20),
```
information about such as publisher, originalPubDate, isbn10, isbn13, etc looks specific to a particular book edition and they have been reflected in the edition table, hence repeating them in the details table looks redundant.

ahchung23 replied:
Each book has a main edition data that is refered to in the details table. All other editions of the book are considered secondary and refered to in the editions table. For goodreads.com the main edition of the book is selected due to a number of factor such as, where the user is accessing the site from and how recently each edition was printed or how popular each edition was. For example, a book that was first published 100 years ago would not show the first published edition as its main edition. It would show the edition that is most readily available for purchase at stores currently.

kazshige commented on database/seedMethods.js:
```diff
@@ -0,0 +1,181 @@
+  //details Table Seeder Function
+  var seedDetailsTable = (details) => {
+    let { type, pageNum, publisher, title, dates, isbn10, isbn13, language } = details;
+    let queryString = 'INSERT INTO details (type, pagenum, publisher, firstPubDate, originalPubDate, title, isbn10, isbn13, language) values (?, ?, ?, ?, ?, ?, ?, ?, ?);';
```
nice attempt, but if the table comments above have been reconciled, you may need to revisit this seeder and ensure that only required fields are seeded

ahchung23 replied:
👍 I think I will take a good look at what information might be redundant. However, I'm pretty sure the isbn numbers, and publication dates, even publishers are all different for every edition.

kazshige commented on server/index.js:
```diff
@@ -0,0 +1,68 @@
+  const id = req.params.id;
+  db.getDetails(id)
+    .then(results => {
+      let details = results[0][0];
```
you may need to verify that results have a length greater than "0" before going ahead to index it in order to prevent the "undefined is not an object error"

ahchung23 replied:
Thank you! I will fix this right away!

kazshige commented on server/index.js:
```diff
@@ -0,0 +1,68 @@
+    })
+    .catch(err => {
+      console.log('err getting extra details!', err);
+      res.send(err);
```
you may also want to send the appropriate status code for error messages for a 200 status code will be parsed as success by the client

ahchung23 replied:
Gotcha. Thanks for the great tip :)

kazshige commented on server/index.js:
```diff
@@ -0,0 +1,68 @@
+
+let staticPath = __dirname + '/../public';
+console.log(staticPath);
+app.use('/books/:id', express.static(staticPath));
```
I believe it will be more appropriate to use the index route "/" to serve the static files. Because the index will serve as the entry point to the application and new users may not be able to know any other route. I will fix mine too.

ahchung23 replied:
I think the "/" route should be theoretically the main goodreads page. The static file we serve should be an HTML page showing ONE book item with a specific id. So the static route should be "/book/:id"

ahchung23 commented:
**Changes to Features:**

-seeds database without using schema.sql!
the database and tables are now created within seed.js using only node (DOES NOT USE schema.sql)
schema.sql only included within repo for clarity of database design.
seed.js creates a new database connection and closes connection once seeding data is finished
package.json npm script changed to reflect this --> "seed" :  "node database/seed.js"

-drop tables made easier by dropTables.js and npm script --> "table:drop" : "drop:tables": "node database/dropTables.js"

-routes get requests now detect if DB returned an empty query for wrong book Id
added else statements to GET routing. Now sends a response with error status code and error messages when the id parameter within the request URL does not match a book within the database!

-repo now includes tests for server routing and endpoints.
tests were written using JEST and SUPERTEST
they are contained within test folder and server.test.js

-repo now uses circleci to implement continuous integration

**Questions:**
Do you see any potential problems with how I'm routing different endpoints in for my API?
Do you see any problems with how I'm seeding my database?

Feedback much appreciated! 👍

ahchung23 commented:
**Changes to Features:**

-seeds database without using schema.sql!
the database and tables are now created within seed.js using only node (DOES NOT USE schema.sql)
schema.sql only included within repo for clarity of database design.
seed.js creates a new database connection and closes connection once seeding data is finished
package.json npm script changed to reflect this --> "seed" :  "node database/seed.js"

-added new table "settings" to database and seeded with data.
settings table includes: (id, city, country bookId)

-drop tables made easier by dropTables.js and npm script --> "table:drop" : "drop:tables": "node database/dropTables.js"

-get requests now detect if DB returned an empty query for wrong book Id
added else statements to GET routing. Now sends a response with error status code and error messages when the id parameter within the request URL does not match a book within the database!

-repo now includes tests for server routing and endpoints.
tests were written using JEST and SUPERTEST
they are contained within test folder and server.test.js

-repo now uses circleci to implement continuous integration

-repo now follows Airbnb style guide with eslint!

**Questions:**
Do you see any potential problems with how I'm routing different endpoints in for my API?
Do you see any problems with how I'm seeding my database?

Feedback much appreciated! 👍 
Ticket:
https://trello.com/c/KSHKgKbn

kazshige commented on server/app.js:
```diff
@@ -0,0 +1,71 @@
+      } else {
+        res.send(details);
+      }
+    })
```
you can try to add a catch method to your promises in case the request gets stuck for any reason, so this will help prevent a connection timeout error

ahchung23 replied:
good point! I will add that right away :)

kazshige commented on database/seed.js:
```diff
@@ -0,0 +1,103 @@
+          bookId INT
+        );`);
+    })
+    .error((err) => {
```
Can you tell me what is the motivation behind creating the db tables in the seed.js file and do you think cascading of promises is a good programming practice I would rather let things run independently?

ahchung23 replied:
I created tables in the seed.js to exclude installing a mysql client within my circleci environment. This made it easier for me to setup my circleci environment. 

If I want to seed my database by running the command mysql -u root < schema.sql within the npm seed command, I needed to install mysql-client within my circleci environment which made setting up the circleci environment harder. It was easier to simply use only a node command since node.js is my top dependancy in my circleci environment.

As for cascading promises, you're right that it would be more efficient to create the tables concurrently using a Promise.all method. However, I think I'll leave that in my backlogs to change later for now since there is not much difference other than taking a little longer during the seeding process when a promise chain is used.

kazshige replied:
Ok, I see. Thanks for answering!
